cs/src/cloud/AzurePageBlobDevice.cs

```diff
@@ -0,0 +1,156 @@
+
+            // Even though Azure Page Blob does not make use of Overlapped, we populate one to conform to the callback API
+            Overlapped ov = new Overlapped(0, 0, IntPtr.Zero, asyncResult);
+            NativeOverlapped* ovNative = ov.UnsafePack(callback, IntPtr.Zero);
+
+            UnmanagedMemoryStream stream = new UnmanagedMemoryStream((byte*)destinationAddress, readLength, readLength, FileAccess.Write);
+
+            // TODO(Tianyu): This implementation seems to swallow exceptions that would otherwise be thrown from the synchronous version of this
+            // function. I wasn't able to find any good documentaiton on how exceptions are propagated or handled in this scenario. 
+            pageBlob.BeginDownloadRangeToStream(stream, (Int64)sourceAddress, readLength, ar => {
+                // Should propagate any exceptions
+                pageBlob.EndDownloadRangeToStream(ar);
+                callback(0, readLength, ovNative);
+            }, asyncResult);
+        }
+        /// <summary>
+        /// <see cref="IDevice.WriteAsync(IntPtr, int, ulong, uint, IOCompletionCallback, IAsyncResult)">Inherited</see>
+        /// </summary>
+        public override unsafe void WriteAsync(IntPtr sourceAddress, int segmentId, ulong destinationAddress, uint numBytesToWrite, IOCompletionCallback callback, IAsyncResult asyncResult)
+        {
+            if (!blobs.TryGetValue(segmentId, out CloudPageBlob pageBlob))
+            {
+
+                // If no blob exists for the segment, we must first create the segment asynchronouly. (Create call takes ~70 ms by measurement)
+                pageBlob = container.GetPageBlobReference(blobName + segmentId);
+
+                // If segment size is -1, which denotes absence, we request the largest possible blob. This is okay because
+                // page blobs are not backed by real pages on creation, and the given size is only a the physical limit of 
+                // how large it can grow to.
+                var size = segmentSize == -1 ? MAX_BLOB_SIZE : segmentSize;
+                // It is up to allocator to ensure that no reads happen before the callback of this function is invoked.
+                if (blobs.TryAdd(segmentId, pageBlob))
+                {
+                    pageBlob.BeginCreate(size, ar => {
+                        pageBlob.EndCreate(ar);
+                        WriteToBlobAsync(pageBlob, sourceAddress, destinationAddress, numBytesToWrite, callback, asyncResult);
+                    }, null);
+                }
+                else
+                {
+                    // Some other thread beat us to calling create, should use their handle to invoke write directly instead
+                    WriteToBlobAsync(blobs[segmentId], sourceAddress, destinationAddress, numBytesToWrite, callback, asyncResult);
```

Comment:
Do we need to somehow delete the page blob reference in this case?

Reply:
I believe it is just a stateless stub that is GCed as references to it are dropped. Again, no way to be sure since I have not found any documentation or specifications stating this.